ARROW-16033: [C++] Pass schema to consuming sink node

westonpace opened the pull request.

westonpace commented:
There is an entirely different implementation we could take (pass the schema in an init method before anything runs):

```
class ARROW_EXPORT SinkNodeConsumer {
 public:
  virtual ~SinkNodeConsumer() = default;
  /// \brief Initialize the consumer with the schema of the batches that will arrive
  virtual Status Init(const std::shared_ptr<Schema>& schema) = 0;
  /// \brief Consume a batch of data
  virtual Status Consume(ExecBatch batch) = 0;
  /// \brief Signal to the consumer that the last batch has been delivered
  ///
  /// The returned future should only finish when all outstanding tasks have completed
  virtual Future<> Finish() = 0;
};
```

Pros:
 * More obvious that the schema will never change
Cons:
 * More methods that have to be implemented
 
 Opinions?  @lidavidm @vibhatha @paleolimbot

lidavidm commented:
I would probably prefer Init/Consume/Finish, we could default-implement Init if it's a concern

vibhatha commented:
This one is cleaner. I also prefer this method for consuming_sink. One question, does this take care of ‘SinkNode’ and ‘OrderBySink’ node too? We may also need to know what’s the schema of the batches we are going to get out of a given sink, am I right? I am merely considering a general case about sink nodes.

vibhatha commented:
Another Advantage is we may be solving this issue too. https://issues.apache.org/jira/browse/ARROW-15297

@westonpace @lidavidm what do you think?

westonpace commented:
> One question, does this take care of ‘SinkNode’ and ‘OrderBySink’ node too?

It doesn't currently fix `SinkNode` or `OrderBySinkNode`.  We could add an init-style schema callback to the options objects for those types.  I'll try it out.

> Another Advantage is we may be solving this issue too. https://issues.apache.org/jira/browse/ARROW-15297

Almost, looks like the python test is failing that was failing on that issue too.  I think we can move towards `WriteNodeOptions` taking either an _optional_ schema (which would replace the one retrieved from the node, we could add a check to make sure they are compatible in the init step) or just a dictionary of metadata that we append to the schema in the init step.

vibhatha commented:
Looks like, if we can address these issues, this PR can cover these grounds. 
@westonpace Thanks a lot for taking your time on this. I am +1 for this modification. 
Shall we also add a few test cases to show the functionality?

westonpace commented:
> I would probably prefer Init/Consume/Finish, we could default-implement Init if it's a concern

I've updated this PR to use an `Init` method.  I did not default-implement as I'm not sure any meaningful consumer can do much without the schema.

I could add a default implementation that actually stored the schema (in a protected variable) but I kind of prefer pure-virtual interfaces for the public API, though maybe that is just my Java/C# background.  Happy to change if anyone has any strong opinion on the matter.

> Another Advantage is we may be solving this issue too. https://issues.apache.org/jira/browse/ARROW-15297

This changes the `schema` to an (optional) `custom_metadata`  which I think means we do indeed solve ARROW-15297.

>  Shall we also add a few test cases to show the functionality?
@vibhatha 

I've added a test to make sure we test the failure path (i.e. `Init` returns a non-ok status).  I'm not sure what other test cases you had in mind.

lidavidm commented on cpp/src/arrow/compute/exec/options.h:
```diff
@@ -304,10 +310,9 @@ class ARROW_EXPORT TableSinkNodeOptions : public ExecNodeOptions {
  public:
   TableSinkNodeOptions(std::shared_ptr<Table>* output_table,
                        std::shared_ptr<Schema> output_schema)
```
We can remove this second argument now right?

westonpace replied:
Good catch.  I've removed it.

vibhatha reviewed (approved):
Looks good to me @westonpace. This will help in exposing Substrait plan very neatly. I cannot think of any additional test cases since, this is basically provides a method to extract the schema in the user code.

westonpace commented:
It seems an R change managed to sneak in that exposes a little dilemma.  The output schema is including the augmented fields, even though the Substrait plan didn't call for them.

I'm still processing how we want to handle this but I'm thinking right now we just make sure not to include augmented fields when running against a Substrait plan.

westonpace commented:
I added an explicit selection step to drop those columns.  I think the proper fix will have to be done in a follow-up.  I added a comment to the R test to this effect.  @paleolimbot mind taking a quick look for sanity's sake?  Then I will merge.

https://github.com/apache/arrow/pull/12721/commits/aad80397f79f542e09360ead42e5e31ecb6259a0
